recipe/2024-04-terraform-provider-secrets.md

```diff
@@ -0,0 +1,259 @@
+
+#### Disadvantages (of each option considered)
+N/A
+
+#### Proposed Option
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+###Recipes (Engine and Driver)
+
+```go
+Note: DataModel in [Design Document to support multiple Terraform Providers](https://github.com/radius-project/design-notes/pull/39/files)
+DataModel in Design doc describes SecretReference:
+type SecretReference {
+  source: string;		// SecretStore Id					
+  key: string;
+}
+
+type RecipeSecretIDs struct {
+  moduleSecret string                  // SecretStoreId associated with git private terraform repo source
+  providerSecrets map[string]map[string]SecretReference  // map[providerName or provider.aliasName][secret_name]SecretReference
+  envSecrets map[string]SecretReference // map[environmentvariablename]SecretReference 
+}
+
+type RecipeSecrets struct {
+  moduleSecrets v20231001preview.SecretStoresClientListSecretsResponse
+  providerSecrets map[string]map[string]string   // map[providerName/provider.aliasName][secret_name]value
+  envSecrets map[string]string                   // map[environmentvariablename]value 
+}
+
+
+// FindSecretIDs will be updated in current DriverWithSecrets interface to now return RecipeSecretIDs
+type DriverWithSecrets interface {
+	// Driver is an interface to implement recipe deployment and recipe resources deletion.
+	Driver
+
+	// FindSecretIDs gets the secret store resource IDs for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (**RecipeSecretIDs**, error)
+}
+
+LoadSecrets in SecretsLoader interface will be updated to LoadRecipeSecrets(), It will internally call functions which will populate resolved secrets based on the Ids in RecipeSecretIDs{}. The result is combined as RecipeSecrets{} which is passed as BaseOptions.Secrets to the Driver in Execute() call.
+
+LoadRecipeSecrets(RecipeSecretIDs, ...)(RecipeSecrets, error){
+  ...
+  loadModuleSecrets()
+  loadProviderSecrets()
+  loadEnvSecrets()
+}
+
+
+// When calling driver.Execute we pass in BaseOptions which is structured as:
+type BaseOptions struct {
+	// Configuration is the configuration for the recipe.
+	Configuration recipes.Configuration
+
+	// Recipe is the recipe metadata.
+	Recipe recipes.ResourceMetadata
+
+	// Definition is the environment definition for the recipe.
+	Definition recipes.EnvironmentDefinition
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets **RecipeSecrets**
+}
+```
+
+### Error Handling
+
+If there is an error when iterating through the recipe configuration or calling ListSecrets API, error will be returned. These errors will not include sensitive data on secret keys or values.
```

Comment:
Will there be any new error types?

Reply:
no new error type has come up yet. Will update this note and document if that changes